Kernel/Syscalls/fallocate.cpp

```diff
@@ -31,13 +33,17 @@ ErrorOr<FlatPtr> Process::sys$posix_fallocate(int fd, Userspace<off_t const*> us
         return EFBIG;
 
     auto description = TRY(open_file_description(fd));
+
+    // [EBADF] The fd argument references a file that was opened without write permission.
     if (!description->is_writable())
         return EBADF;
 
+    // [ESPIPE] The fd argument is associated with a pipe or FIFO.
     if (description->is_fifo())
         return ESPIPE;
 
-    if (!S_ISREG(TRY(description->file().stat()).st_mode))
+    // [ENODEV] The fd argument does not refer to a regular file.
+    if (!description->file().is_regular_file())
         return ENODEV;
 
     VERIFY(description->file().is_inode());
```

Comment:
The reason why the separate `is_inode` check is there to begin with is because we are relying on the raw inode just below that (as we don't have a generic truncate implementation).

Depending on the implementation of `is_regular_file` the `VERIFY` indeed won't have any effect, but logically it still makes sense (and, in my opinion, should be kept).

Reply:
More discussion on this happened in #14326.